### PR TITLE
import cdk-validium-node genesis improvement

### LINKED ISSUE #3778: Synchronization of L1InfoTree events before genesis allow to resume
# Rationale

If zkEVM rollup is deployed aftet LxLy upgrade need to build the L1InfoTree
This is corresponding to this [PR](https://github.com/0xPolygon/cdk-validium-node/pull/153) from cdk-validium-node

### PR DESCRIPTION
**Pull request by joanestebanr**
Closes #3778, #3779

### What does this PR do?

Add param `rollupManagerCreationBlockNumber` in genesis file: 
```
    "root": "0x3f86b09b43e3e49a41fc20a07579b79eba044253367817d5c241d23c0e2bc5c9",
    "genesisBlockNumber": 16896721,
    "rollupManagerCreationBlockNumber": 16896721,
    "genesis": [
```

### Reviewers

Main reviewers:

- @John
- @Doe

Codeowner reviewers:

- @-Alice
- @-Bob

### CODE REVIEW

**Review by agnusmor — APPROVED** [review]

File: synchronizer/synchronizer.go [hunk 1]
```diff
@@ -278,6 +279,162 @@ func rollback(ctx context.Context, dbTx pgx.Tx, err error) error {
...
+	log.Warnf("Genesis block not processed yet. Last block synced: %d < genesis %d", lastEthBlockSynced.BlockNumber, s.genesis.BlockNumber)
+	return false, lastEthBlockSynced, nil
+}
+
+// getStartingL1Block find if need to update and if yes the starting point:
+// bool -> need to process blocks
+// uint64 -> first block to synchronize
+// error -> error
+// 1. First try to get last block on DB, if there are could be fully synced or pending blocks
+// 2. If DB is empty the LxLy upgrade block as starting point
```
Comment (ARR552) [inline synchronizer/synchronizer.go]: Can you rephrase this?

File: etherman/etherman.go [hunk 2]
```diff
@@ -41,6 +41,11 @@ import (
 	"golang.org/x/crypto/sha3"
 )
 
+const (
+	// ETrogUpgradeVersion is the version of the LxLy upgrade
+	ETrogUpgradeVersion = 2
```
Comment (ARR552) [inline etherman/etherman.go]: What `ETrogUpgradeVersion = 2` means?

File: etherman/etherman.go [hunk 3]
```diff
@@ -339,6 +344,26 @@ func (etherMan *Client) VerifyGenBlockNumber(ctx context.Context, genBlockNumber
 	return true, nil
 }
 
+// GetL1BlockUpgradeLxLy It returns the block genesis for LxLy before genesisBlock or error
+func (etherMan *Client) GetL1BlockUpgradeLxLy(ctx context.Context, genesisBlock uint64) (uint64, error) {
+	it, err := etherMan.RollupManager.FilterInitialized(&bind.FilterOpts{
+		Start:   1,
+		End:     &genesisBlock,
+		Context: ctx,
+	})
```
Comment (ARR552) [inline etherman/etherman.go]: This won't work in many providers. You must do the query in chunks

File: state/block.go [hunk 4]
```diff
@@ -15,6 +16,10 @@ type Block struct {
 	Checked     bool
 }
 
+func (b *Block) String() string {
+	return fmt.Sprintf("BlockNumber: %d, BlockHash: %s, ParentHash: %s, ReceivedAt: %s", b.BlockNumber, b.BlockHash, b.ParentHash, b.ReceivedAt)
+}
+
```
Comment (ARR552) [inline state/block.go:22]: I think this is unused code. please add it when it is necessary

File: synchronizer/synchronizer_test.go [hunk 5]
```diff
@@ -176,7 +183,7 @@ func TestForcedBatchEtrog(t *testing.T) {
 			m.State.
 				On("GetLastBlock", ctx, m.DbTx).
 				Return(lastBlock0, nil).
-				Once()
+				Maybe()
```
Comment (ARR552) [inline synchronizer/synchronizer_test.go]: This is not very accurate

File: synchronizer/synchronizer_test.go [hunk 6]
```diff
@@ -996,7 +1020,7 @@ func TestReorg(t *testing.T) {
 			m.State.
 				On("GetLastBlock", ctx, m.DbTx).
 				Return(lastBlock1, nil).
-				Once()
+				Maybe()
```
Comment (ARR552) [inline synchronizer/synchronizer_test.go]: same. You use it in many places. Please remove all maybe. You don't know what the code is doing using a maybe tag

**Review by ARR552 — CHANGES_REQUESTED** [review]

File: synchronizer/synchronizer.go (hunk 1 above)
Comment (joanestebanr) [reply synchronizer/synchronizer.go]: I have rewrite as: 
```
// 1. Check last synced block on DB, if there are any could be fully synced (>=genesis) or syncing pre-genesis events (<genesis)
// 2. If no block DB then get the LxLy upgrade block as starting point for pre-genesis or use the genesis block as starting point
```

File: etherman/etherman.go (hunk 2 above)
Comment (joanestebanr) [reply etherman/etherman.go]: This the way that @invocamanman   suggested to discover the block: check the event `initialized` where  `version` is `ETrogUpgradeVersion`

File: state/block.go (hunk 4 above)
Comment (joanestebanr) [reply state/block.go:22]: This code is used in `synchronizer.go:854`

File: synchronizer/synchronizer_test.go (hunk 5 above)
Comment (joanestebanr) [reply synchronizer/synchronizer_test.go]: I remove the sentence

File: synchronizer/synchronizer_test.go (hunk 6 above)
Comment (joanestebanr) [reply synchronizer/synchronizer_test.go]: I have removed all unused Maybe() senteces

File: etherman/etherman.go (hunk 3 above)
Comment (joanestebanr) [reply etherman/etherman.go]: I have checked with Alchemy and Infura..  which other provider do you think that I have to test it?

File: etherman/etherman.go (hunk 2 above)
Comment (jesusholoo) [reply etherman/etherman.go]: I guess this is the initialize version!?, i would even specify something like:
`initializeRollupManagerEtrogVersion` or something like that ^^

File: etherman/etherman.go (hunk 3 above)
Comment (joanestebanr) [reply etherman/etherman.go]: I have add a param on genesis to by-pass the auto-discover if the provider doesn't support this

File: etherman/etherman.go (hunk 2 above)
Comment (joanestebanr) [reply etherman/etherman.go]: Finally I change the implementation in favour of L1InfoTreeUpdate that have more sense
